### PR TITLE
Prevent deadlock after an unmanaged error in MPI_SENDRECV

### LINKED ISSUE #9160: ibm/pt2pt/sendrecv_big test hangs when using OFI transport and some OFI providers
The ibm pt2pt sendrecv_big test hangs when using the OFI MTL and PSM2 libfabric provider, and other providers with a maximum fi_tsend message length.  

The test emits the expected warning message:
Message size 30720000000 bigger than supported by selected transport. Max = 4294963200
and then hangs.  

The OFI MTL is properly returning an error code to the upper layers of OMPI, but this code in sendrecv.c
```
 #if OPAL_ENABLE_FT_MPI
         /* If ULFM is enabled we need to wait for the posted receive to
          * complete, hence we cannot return here */
         rcs = rc;
 #else
```
results in the error return code being ignored.

If one comments out these OPAL_ENABLE_FT_MPI code blocks the test errors out as would be expected:
```
[x232:00000] *** An error occurred in MPI_Sendrecv
[x232:00000] *** reported by process [2952003585,0]
[x232:00000] *** on communicator MPI_COMM_SELF
[x232:00000] *** MPI_ERR_OTHER: known error not in list
[x232:00000] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[x232:00000] ***    and MPI will try to terminate your MPI job as well)
```

this problem is observed on master and v5.0.x branches.

### PR DESCRIPTION
**Pull request by abouteiller**
fixes #9160

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

### CODE REVIEW

**Comment by abouteiller** [comment]
@hppritcha I don't have PSM2 hardware, can you check if that solves the original report?

**Comment by jsquyres** [comment]
@hppritcha is out on vacation.  @mwheinz can you guys check?

**Comment by mwheinz** [comment]
> @hppritcha is out on vacation. @mwheinz can you guys check?

Sorry, @jsquyres - I was on vacation as well. Do we have a reproducer for the problem?

**Comment by bosilca** [comment]
Yes, check #9160 for the experimental setup that triggered this issue.

**Comment by mwheinz** [comment]
@abouteiller - I don't think the problem is your patch, but when I try to compile your branch of OMPI I'm getting link errors:

```
//usr/mpi/gcc/openmpi-expr/lib/libopen-pal.so.0: undefined reference to `ucm_test_external_events'
//usr/mpi/gcc/openmpi-expr/lib/libopen-pal.so.0: undefined reference to `uct_query_components'
//usr/mpi/gcc/openmpi-expr/lib/libopen-pal.so.0: undefined reference to `uct_release_component_list'
/usr/mpi/gcc/openmpi-expr/lib/libmpi.so: undefined reference to `ucp_tag_recv_nbx'
//usr/mpi/gcc/openmpi-expr/lib/libopen-pal.so.0: undefined reference to `uct_component_query'
/usr/mpi/gcc/openmpi-expr/lib/libmpi.so: undefined reference to `ucp_tag_send_nbx'
```
I'm building with:
```
(./autogen.pl && ./configure --prefix=/usr/mpi/gcc/openmpi-expr --with-hwloc=internal --with-libevent=internal --with-pmix=internal --with-psm2 && make -j 32) 2>&1 | tee make.log
```

Any thoughts

**Review by mwheinz — APPROVED** [review]
Patch appears to fix the hang.

File: ompi/mca/pml/ob1/pml_ob1_recvfrag.c [hunk 1]
```diff
@@ -681,7 +681,7 @@ void mca_pml_ob1_recv_frag_callback_ack (mca_btl_base_module_t *btl,
 #if OPAL_ENABLE_FT_MPI
     /* if the req_recv is NULL, the comm has been revoked at the receiver */
```
Comment (hppritcha) [inline ompi/mca/pml/ob1/pml_ob1_recvfrag.c:682]: kind of unusual to not have a missing  endif

File: ompi/mpi/c/sendrecv.c [hunk 2]
```diff
@@ -92,21 +92,27 @@ int MPI_Sendrecv(const void *sendbuf, int sendcount, MPI_Datatype sendtype,
...
+            if (OPAL_UNLIKELY(MPI_ERR_PROC_FAILED != rc))
+            /* if intentionally spills outside ifdef */
+#endif
+            ompi_request_cancel(req);
+        }
     }
 
     if (source != MPI_PROC_NULL) { /* wait for recv */
         rc = ompi_request_wait(&req, status);
 #if OPAL_ENABLE_FT_MPI
```
Comment (hppritcha) [inline ompi/mpi/c/sendrecv.c:112]: don't we need a matching endif ?

Comment (bosilca) [reply ompi/mpi/c/sendrecv.c:112]: line 115 below. It does not have the comment to indicate what it is matched with, but it's there.

Comment (hppritcha) [reply ompi/mpi/c/sendrecv.c:112]: hmmph.  i do need new glasses

**Review by hppritcha — APPROVED** [review]

**Pull request merged**